### PR TITLE
fix: Two small fixes in background tasks

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-596

All tests will pass in the final [PR](https://github.com/keboola/keboola-as-code/pull/1770).

[Large set of fixes and E2e tests](https://github.com/keboola/keboola-as-code/pull/1759) is divided into smaller PRs to do a review.

--------------

**Changes:**
- Task prepare phase is now connected to the caller context - fixed telemetry.
- The Stream API is using `StartTask`, instead of `RunTask`.
  - There was a bug, tasks were running in the foreground and were blocking the completion of the request.

-----------

Commend per commits. Please, explore [the final version](https://github.com/keboola/keboola-as-code/pull/1759) in the IDE.

-------------

### CODE REVIEW

File: internal/pkg/service/common/task/node.go [hunk 1]
```diff
@@ -168,15 +164,15 @@ func (n *Node) StartTask(ctx context.Context, cfg Config) (t Task, err error) {
...
 }
 
 // RunTask in foreground, the task run at most once, it is provided by local lock and etcd transaction.
-func (n *Node) RunTask(cfg Config) (t Task, err error) {
+func (n *Node) RunTask(ctx context.Context, cfg Config) (t Task, err error) {
 	// Prepare task, acquire lock, handle error during prepare phase
-	task, fn, err := n.prepareTask(n.tasksCtx, cfg)
+	task, fn, err := n.prepareTask(ctx, cfg)
 	if err != nil {
 		return Task{}, err
```
Comment (michaljurecko) [inline internal/pkg/service/common/task/node.go:177]: The prepare task phase is now using caller context, so it is now correctly reported by the telemetry.

Comment (michaljurecko) [reply internal/pkg/service/common/task/node.go:177]: Task run phase has its own background context, see next commit.

File: internal/pkg/service/stream/api/service/task.go [hunk 2]
```diff
@@ -43,13 +45,19 @@ func (s *service) startTask(cfg taskConfig) (task.Task, error) {
 	}
 
 	taskID := task.ID(cfg.Type + "/" + objectKey)
-	return s.tasks.RunTask(task.Config{
+	return s.tasks.StartTask(ctx, task.Config{
 		Type: cfg.Type,
 		Key:  task.Key{ProjectID: cfg.ProjectID, TaskID: taskID},
 		Lock: cfg.ObjectKey.String(),
```
Comment (michaljurecko) [inline internal/pkg/service/stream/api/service/task.go:51]: Fixed bug in API, `RunTask` -> `StartTask`.

File: internal/pkg/service/stream/api/service/task.go [hunk 3]
```diff
@@ -43,13 +45,19 @@ func (s *service) startTask(cfg taskConfig) (task.Task, error) {
...
 		Context: func() (context.Context, context.CancelFunc) {
-			return context.WithTimeout(context.Background(), cfg.Timeout)
+			return context.WithTimeout(context.WithoutCancel(ctx), cfg.Timeout)
+		},
```
Comment (michaljurecko) [inline internal/pkg/service/stream/api/service/task.go:54]: Here is factory of the task background context, we can use relatively new `context.WithoutCancel` call, I will fix it also in the other places.

File: internal/pkg/service/stream/api/service/task.go [hunk 4]
```diff
@@ -43,13 +45,19 @@ func (s *service) startTask(cfg taskConfig) (task.Task, error) {
...
+		Operation: func(ctx context.Context, logger log.Logger) task.Result {
+			rb := rollback.New(logger)
+			ctx = rollback.ContextWith(ctx, rb)
+			result := cfg.Operation(ctx, logger)
+			rb.InvokeIfErr(ctx, &result.Error)
+			return result
```
Comment (michaljurecko) [inline internal/pkg/service/stream/api/service/task.go:60]: Keboola table sink bridge operations need `rollback` instance in the context.

**Review by hosekpeter — APPROVED** [review]
LGTM

**Review by Matovidlo — APPROVED** [review]
LGTM

**Pull request merged**